<launch>
    <arg name="load_manager" default="true"/>
    <arg name="manager_name" default="camera_manager"/>
    <arg name="manager_threads" default="4"/>
    <arg name="camera_name" default="cam0"/>

    <group ns="cams/$(arg camera_name)">
        <node if="$(arg load_manager)" pkg="nodelet" type="nodelet" name="$(arg manager_name)" args="manager" output="screen">
            <param name="num_worker_threads" value="$(arg manager_threads)"/>
        </node>

        <!-- De-Bayer / rectify -->
        <include file="$(find image_proc)/launch/image_proc.launch">
            <arg name="manager" value="$(arg manager_name)"/>
        </include>

        <!-- Apriltag detection -->
        <node pkg="apriltag_ros" type="apriltag_ros_continuous_node" name="apriltag" output="screen">
            <param name="camera_frame" type="str" value="$(arg camera_name)"/>
            <param name="publish_tag_detections_image" type="bool" value="true"/>
            <rosparam command="load" file="$(find rktl_perception)/config/apriltag_settings.yaml"/>
            <rosparam command="load" file="$(find rktl_perception)/config/tags.yaml"/>
        </node>

        <!-- Localization -->
        <node pkg="rktl_perception" type="localizer" name="localizer" output="screen">
            <param name="origin_id" type="str" value="0"/>
            <param name="buffer_size" type="int" value="300"/>
            <param name="ball_sub_topic" type="str" value="ball_vec"/>
            <param name="ball_pub_topic" type="str" value="/ball/pose"/>
            <param name="ball_radius" type="double" value="0.05"/>
            <rosparam param="pub_topics">
          '0': /origin/pose
          '2': /cars/car0/pose
            </rosparam>
        </node>

        <!-- Ball tracking -->
        <node pkg="rktl_perception" type="ball_detection" name="ball_detection">
            <param name="publishThresh" value="true"/>
            <param name="min_hue" value="50"/>
            <param name="min_sat" value="50"/>
            <param name="min_vib" value="50"/>
            <param name="max_hue" value="80"/>
            <param name="max_sat" value="255"/>
            <param name="max_vib" value="210"/>
        </node>
        <!-- pointgrey RGB camera nodelet -->
        <node pkg="nodelet" type="nodelet" name="$(arg camera_name)" args="load pointgrey_camera_driver/PointGreyCameraNodelet $(arg manager_name)" output="screen">
            <rosparam command="load" file="$(find rktl_perception)/config/$(arg camera_name)/pointgrey.yaml"/>
            <param name="frame_id" value="$(arg camera_name)"/>
            <param name="camera_info_url" value="package://rktl_perception/config/$(arg camera_name)/calibration.yaml"/>
        </node>
    </group>
</launch>
